### PR TITLE
Add download bytecode button to explorer

### PR DESCRIPTION
**Pull request by losman0s**
#### Problem
Working towards easing on-chain program inspectability/verifyability, it might be useful to be able to download a program's bytecode directly from the Explorer.

#### Summary of Changes
Add a "download bytecode" button in the account details view for program accounts.

### CODE REVIEW

File: explorer/src/components/account/UpgradeableLoaderAccountSection.tsx [hunk 1]
```diff
@@ -179,7 +180,16 @@ export function UpgradeableProgramDataSection({
         {account.details?.space !== undefined && (
           <tr>
             <td>Data (Bytes)</td>
-            <td className="text-lg-end">{account.details.space}</td>
+            <td className="text-lg-end">
+              <Downloadable
+                data={programData.data[0]}
```
Comment (jstarry) [inline explorer/src/components/account/UpgradeableLoaderAccountSection.tsx:185]: I'd prefer if this downloaded binary data instead of a base64 string

File: explorer/src/components/account/UpgradeableLoaderAccountSection.tsx [hunk 2]
```diff
@@ -179,7 +180,16 @@ export function UpgradeableProgramDataSection({
...
+                filename={`bytecode-${account.pubkey
+                  .toString()
+                  .slice(0, 6)}.txt`}
```
Comment (jstarry) [inline explorer/src/components/account/UpgradeableLoaderAccountSection.tsx:188]: Prefer to have the full key in the filename

**Review by jstarry — COMMENTED** [review]
Looking good! Thanks for taking this on

File: explorer/src/components/account/UpgradeableLoaderAccountSection.tsx (hunk 1 above)
Comment (losman0s) [reply explorer/src/components/account/UpgradeableLoaderAccountSection.tsx:185]: Conversion done in the `Downloadable` component. Two observations:

1. `Downloadable` is probably early, halfway optimization but I did not want to:
* pollute `UpgradeableLoaderAccountSection`, nor
* needlessy make `Downloadable` as generic as it should be (since it's a new otherwise unused component anyway)
2. the downloaded data differs by a few leading bytes from what gets downloaded through `solana account`. I'm not 100% sure what part of the data that is, but I'm guessing the generic attributes that already get parsed? (`lamports`/`executable`/etc)

Let me know if you have issue with these points or if this will do.

Comment (jstarry) [reply explorer/src/components/account/UpgradeableLoaderAccountSection.tsx:185]: The difference is due to the [45 bytes of metadata](https://github.com/solana-labs/solana/blob/3bd5a89d6ff680bdfc2a64b30081ecd46548d19a/sdk/program/src/bpf_loader_upgradeable.rs#L286) which is stored before the executable in the account. When using `solana account`, the full raw data of an account is downloaded.

```
> solana account 99hNSg9acmyg5hWTjw76GoA8k3LG5NvdPXih7sotdmZ7 -o 99hNSg9acmyg5hWTjw76GoA8k3LG5NvdPXih7sotdmZ7.bin --url testnet
> ls -l
total 5808
-rw-r--r--  1 starry  staff  1485581 Jan 16 11:49 99hNSg9acmyg5hWTjw76GoA8k3LG5NvdPXih7sotdmZ7.bin
-rw-r--r--@ 1 starry  staff  1485536 Jan 16 11:50 bytecode-99hNSg9acmyg5hWTjw76GoA8k3LG5NvdPXih7sotdmZ7.txt
```

The `data` field of `programData` is populated by RPC using this account parser: https://github.com/solana-labs/solana/blob/3bd5a89d6ff680bdfc2a64b30081ecd46548d19a/account-decoder/src/parse_bpf_loader.rs#L10 and you can see that it takes into account the "offset" of where the executable data actually starts.

Long story short, your change is correct! And also, doing the conversion in `Downloadable` looks good to me.

File: explorer/src/components/account/UpgradeableLoaderAccountSection.tsx [hunk 3]
```diff
@@ -179,7 +180,14 @@ export function UpgradeableProgramDataSection({
         {account.details?.space !== undefined && (
           <tr>
             <td>Data (Bytes)</td>
-            <td className="text-lg-end">{account.details.space}</td>
+            <td className="text-lg-end">
+              <Downloadable
+                data={programData.data[0]}
+                filename={`bytecode-${account.pubkey.toString()}.txt`}
```
Comment (jstarry) [inline explorer/src/components/account/UpgradeableLoaderAccountSection.tsx:186]: another nit: Could you change the extension to `.bin` to indicate that the file is binary data now? I sort of prefer to drop the "bytecode-" prefix as well but not a big deal, choose whatever you think is clearest.

Comment (losman0s) [reply explorer/src/components/account/UpgradeableLoaderAccountSection.tsx]: Extension changed and prefix dropped, it did not bring much.

**Review by jstarry — APPROVED** [review]

**Comment by jstarry** [comment]
Thanks for the contribution @losman0s!

**Comment by t-nelson** [comment]
Is this being served directly from RPC and are we concerned about a potential increase in load?

**Comment by losman0s** [comment]
> Is this being served directly from RPC and are we concerned about a potential increase in load?

This is making use of the account data already fetched upon page load, and merely making is available to download through the button. So no additional RPC call and no extra load incurred.

**Comment by t-nelson** [comment]
Nice!  Thanks for the clarification!
